internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/server/VeniceServerTest.java

```diff
@@ -182,6 +183,37 @@ public void testCheckBeforeJointClusterBeforeHelixInitializingCluster() throws E
     }
   }
 
+  @Test
+  public void testStartServerAndShutdownWithPartitionAssignmentVerification() {
+    try (VeniceClusterWrapper cluster = ServiceFactory.getVeniceCluster(1, 0, 0)) {
+      Properties featureProperties = new Properties();
+      featureProperties.setProperty(SERVER_ENABLE_SERVER_ALLOW_LIST, Boolean.toString(true));
+      featureProperties.setProperty(SERVER_IS_AUTO_JOIN, Boolean.toString(true));
+      cluster.addVeniceServer(featureProperties, new Properties());
+      VeniceServerWrapper server = cluster.getVeniceServers().get(0);
+      Assert.assertTrue(server.getVeniceServer().isStarted());
+      StorageService storageService = server.getVeniceServer().getStorageService();
+      StorageEngineRepository repository = storageService.getStorageEngineRepository();
+      Assert
+          .assertTrue(repository.getAllLocalStorageEngines().isEmpty(), "New node should not have any storage engine.");
+
+      // Create a storage engine.
+      String storeName = Version.composeKafkaTopic(cluster.createStore(1), 1);
+      Assert.assertEquals(repository.getAllLocalStorageEngines().size(), 1);
+      Assert.assertTrue(server.getVeniceServer().getHelixParticipationService().isRunning());
+      Assert.assertEquals(storageService.getStorageEngine(storeName).getPartitionIds().size(), 3);
+
+      server.getVeniceServer().shutdown();
+      cluster.getControllerClient().deleteStore(storeName);
+
+      cluster.stopVeniceServer(server.getPort());
+      cluster.restartVeniceServer(server.getPort());
+      repository = server.getVeniceServer().getStorageService().getStorageEngineRepository();
+      Assert.assertEquals(repository.getAllLocalStorageEngines().size(), 1);
```

Comment:
If we've dropped all partitions, I think we should delete the entire store.

Reply:
Re-opening this one.

In our case, we're deleting the whole store which causes us to delete all of the partitions. So the `storageEngine` exists with 0 partitions. If this happens, we should drop the `storageEngine` too.

So `repository.getAllLocalStorageEngines().size()` should return 0.

Reply:
Here's an [example](https://github.com/linkedin/venice/blob/78de20c7014f8dc1b86e3e1bb0d165fcb0e06911/clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java#L391-L409). If `remainingPartitions` is empty, it calls `removeStorageEngine`.

To prevent duplicate logic you could just call `dropStorePartition` and set `removeEmptyStorageEngine` to `true`.

Reply:
The storage engine has been dropped upon removal of partitions.